lib/lookup/src/lookup_buf/mod.rs

```diff
@@ -0,0 +1,394 @@
+        Box::new(
+            self.segments
+                .shrink()
+                .filter(|segments| !segments.is_empty())
+                .map(|segments| Self { segments }),
+        )
+    }
+}
+
+impl Display for LookupBuf {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        let mut peeker = self.segments.iter().peekable();
+        let mut next = peeker.next();
+        let mut maybe_next = peeker.peek();
+        while let Some(segment) = next {
+            match (segment, maybe_next) {
+                (SegmentBuf::Field(_), Some(next)) if next.is_field() || next.is_coalesce() => {
+                    write!(f, r#"{}."#, segment)?
+                }
+                (SegmentBuf::Field(_), _) => write!(f, "{}", segment)?,
+                (SegmentBuf::Index(_), Some(next)) if next.is_field() || next.is_coalesce() => {
+                    write!(f, r#"[{}]."#, segment)?
+                }
+                (SegmentBuf::Index(_), _) => write!(f, "[{}]", segment)?,
+                (SegmentBuf::Coalesce(_), Some(next)) if next.is_field() || next.is_coalesce() => {
+                    write!(f, r#"{}."#, segment)?
+                }
+                (SegmentBuf::Coalesce(_), _) => write!(f, "{}", segment)?,
+            }
+            next = peeker.next();
+            maybe_next = peeker.peek();
+        }
+        Ok(())
+    }
+}
+
+impl LookupBuf {
+    /// Creates an lookup to the root
+    pub fn root() -> Self {
+        Self {
+            segments: VecDeque::new(),
+        }
+    }
+
+    pub fn iter(&self) -> std::collections::vec_deque::Iter<'_, SegmentBuf> {
+        self.segments.iter()
+    }
+
+    pub fn to_lookup(&self) -> Lookup {
+        Lookup::from(self)
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.segments.is_empty()
+    }
+
+    pub fn from_segments(segments: Vec<SegmentBuf>) -> Self {
+        Self {
+            segments: segments.into_iter().collect(),
+        }
+    }
+
+    /// Return a borrow of the SegmentBuf set.
+    pub fn as_segments(&self) -> &VecDeque<SegmentBuf> {
+        &self.segments
+    }
+
+    /// Return the SegmentBuf set.
+    pub fn into_segments(self) -> VecDeque<SegmentBuf> {
+        self.segments
+    }
+
+    /// Create the possible fields that can be followed by this lookup.
+    /// Because of coalesced paths there can be a number of different combinations.
+    /// There is the potential for this function to create a vast number of different
+    /// combinations if there are multiple coalesced segments in a path.
+    ///
+    /// The limit specifies the limit of the path depth we are interested in.
+    /// Metrics is only interested in fields that are up to 3 levels deep (2 levels + 1 to check it
+    /// terminates).
+    ///
+    /// eg, .tags.nork.noog will never be an accepted path so we don't need to spend the time
+    /// collecting it.
+    pub fn to_alternative_components(&self, limit: usize) -> Vec<Vec<&str>> {
```

Comment:
This is currently only used to match against the metric fields -> `.name`, `.namespace`, `.tags`, `.tags.<tagname>`, and so on. Since the maximum depth the metric fields can go to, we are only interested in a depth of three (the third level only to ensure a segment hasn't been specified, so someone isn't trying to access say, `.tags.host.thing`.) I'll add a const.